### ISSUE TITLE
Directory symlinks are not followed

### ISSUE THREAD

**Issue by julienw** [eslint/eslint#5427]
While working on the gobble eslint plugin (https://github.com/julienw/gobble-eslint) I realized that eslint is not following symlinks for directories. Note it follows symlinks just fine for normal files.

**Comment by julienw** [comment]
I think this will be improved in https://github.com/eslint/eslint/issues/4606 but from what I understand in that bug, following symlinks is what we should do now.

**Comment by eslintbot** [comment]
Hi @julienw, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by julienw** [comment]
The bug is present in latest eslint v2.2. It's difficult to propose a ready-to-use testcase because the issue is with symbolic links but it should be quite easy to do a local one.

**Comment by ilyavolodin** [comment]
@julienw Do you know if this was working prior to v2.2? It looks like it probably never worked before, due to missing flag for `glob`. I vaguely remember having an issue a while back requesting we remove ability to follow symlinks, maybe that's why it wasn't present?

**Comment by julienw** [comment]
I'm sure it was not working in v2.0, I can check for earlier versions if it's important but I think like you that it never worked :)

**Comment by nzakas** [comment]
@julienw can you please outline a test case that we can reproduce?

**Comment by julienw** [comment]
OK, I'll do a script that can create the environment to reproduce. (not before tomorrow though)

**Comment by julienw** [comment]
@nzakas I created a gist in https://gist.github.com/julienw/face3699c34446c92c52 containing a single shell script.

STR:
1. Run this shell script in an empty directory.
This will create a directory `real` with faulty files: one directly in the directory, one in a subdirectory. And you'll have a link to the directory, inside the `linked` directory.
2. run eslint on `.` and/or on each of the directories `real` and `linked`.

Notice the file `file2.js` is properly checked in all situations, but the file `subdir/file1.js` is not checked when using the symbolic link. This is fixed with my patch. 

Hope this helps !

**Comment by julienw** [comment]
Maybe `file2.js` is checked in all situations whereas it shouldn't because `follow: false` is broken in node-glob: https://github.com/isaacs/node-glob/pull/242 ?

Also I checked the issue also happens in v1.10.

**Comment by nzakas** [comment]
Sorry, lost track of this. Is this still an issue after 2.0?

**Comment by alberto** [comment]
@julienw can you confirm it's still an issue in latest version?

**Comment by julienw** [comment]
Yes it is :)

And adding `follow: true` in https://github.com/eslint/eslint/blob/dfbc112bb22a97449c2b50ad33a3ef14afd7f5ce/lib/util/glob-util.js#L114 fixes it for me. Good that cli-engine.js now uses it :)

(my old patch in https://github.com/eslint/eslint/pull/5422 doesn't apply anymore -- and I still don't have time to work on an automatic test for this, sorry)

**Comment by julienw** [comment]
I tried in v3.2.2.

**Comment by kaicataldo** [comment]
@elsint/eslint-team Looks like this still might be an issue - do we want to accept this or close it?

**Comment by nzakas** [comment]
Yeah, let's accept this. Seems like an easy fix.

**Comment by kaicataldo** [comment]
@julienw Any interest in making a PR for this change?

**Comment by julienw** [comment]
Interest: yes; but time: not at the moment :(

**Comment by jpmarra** [comment]
Is this still an issue?

**Comment by julienw** [comment]
Pretty sure, yes, as `follow: true` is not in https://github.com/eslint/eslint/blob/465e615188923577e6ca88c559f0c085eb0a2b37/lib/util/glob-util.js#L226-L230

I still don't have time to work on a test for this.

**Comment by msikma** [comment]
It's still an issue. I'm on Eslint v4.19.1. However, going to `glob-util.js` and adding `follow: true` does not seem to fix the problem for me.

Here's my [Eslint output](https://gist.github.com/msikma/f58252d639bcfa7956fc48d37a12b059). It can't find my symlinked packages. The project root and working directory is the `ia.js` dir. The packages are correctly symlinked, though:

```bash
$ head ./packages/ia-cli/node_modules/ia-client/package.json 
{
  "name": "ia-client",
  "version": "1.0.0",
  "main": "dist/index.js",
  "homepage": "https://github.com/msikma/ia.js",
  "scripts": {
    "lint": "eslint src; exit 0",
    "build": "babel src --presets babel-preset-env --plugins transform-class-properties,transform-object-rest-spread --out-dir dist"
  },
```

Each package has its own `node_modules` directory with symlinks to the same directories that Eslint complains about, so it's not following the symlinks for me even with `follow: true`.

**Comment by kaicataldo** [comment]
PRs are welcome! I think the only tricky thing here is figuring out how to test it.

**Comment by msikma** [comment]
If no one more experienced has time, I'll have a look at it this weekend to see if I can figure something out. I'm also not entirely sure what kind of test will be sufficient, but I'll give it a try.

**Comment by kaicataldo** [comment]
Sounds great!

> If no one more experienced has time, I'll have a look at it this weekend to see if I can figure something out.

Once you do open source long enough you realize that being willing to take the time is actually the most valuable attribute a contributor can have :)

**Comment by jim-lake** [comment]
Still hitting this problem, especially annoying with npm link
